MapboxNavigation/SpeechSynthesizersController.swift

```diff
@@ -0,0 +1,116 @@
+/// Can be initialized with array of synthesizers which will be called in order of appearance, untill one of them is capable to vocalize current `SpokenInstruction`
+open class SpeechSynthesizersController: SpeechSynthesizing {
+    
+    // MARK: - Properties
+    
+    public var delegate: SpeechSynthesizingDelegate?
+    
+    public var muted: Bool = false {
+        didSet {
+            speechSynthesizers.forEach { $0.muted = muted }
+        }
+    }
+    public var volume: Float = 1.0 {
+        didSet {
+            speechSynthesizers.forEach { $0.volume = volume }
+        }
+    }
+    public var isSpeaking: Bool {
+        return speechSynthesizers.first(where: { $0.isSpeaking }) != nil
+    }
+    public var locale: Locale = Locale.autoupdatingCurrent {
+        didSet {
+            speechSynthesizers.forEach { $0.locale = locale }
+        }
+    }
+    
+    public  let speechSynthesizers: [SpeechSynthesizing]
+    
+    private var currentLegProgress: RouteLegProgress?
+    
+    // MARK: - Lifecycle
+    
+    public init(_ speechSynthesizers: [SpeechSynthesizing]? = nil, accessToken: String? = nil) {
+        self.speechSynthesizers = speechSynthesizers ?? [
+            MapboxSpeechSynthesizer(accessToken),
+            SystemSpeechSynthesizer()
+        ]
+        
+        self.speechSynthesizers.forEach { $0.delegate = self }
+    }
+    
+    // MARK: - Public Methods
+    
+    public func changedIncomingSpokenInstructions(_ instructions: [SpokenInstruction]) {
+        speechSynthesizers.forEach { $0.changedIncomingSpokenInstructions(instructions) }
+    }
+    
+    public func speak(_ instruction: SpokenInstruction, during legProgress: RouteLegProgress) {        
+        guard let synthesizer = speechSynthesizers.first else {
+            assert(false, "SpeechSynthesizersController has 0 speechSynthesizers")
+            delegate?.speechSynthesizer(self,
+                                        didSpeak: instruction,
+                                        with: nil)
+            return
+        }
+        
+        currentLegProgress = legProgress
+        synthesizer.speak(instruction, during: legProgress)
+    }
+    
+    public func stopSpeaking() {
+        speechSynthesizers.forEach { $0.stopSpeaking() }
+    }
+    
+    public func interruptSpeaking() {
+        speechSynthesizers.forEach { $0.interruptSpeaking() }
+    }
+}
+
+extension SpeechSynthesizersController: SpeechSynthesizingDelegate {
+    
+    public func speechSynthesizer(_ speechSynthesizer: SpeechSynthesizing, didSpeak instruction: SpokenInstruction, with error: SpeechError?) {
+        if let error = error {
+            
+            
+            if let index = speechSynthesizers.firstIndex(where: { $0 === speechSynthesizer }),
+                let legProgress = currentLegProgress,
+                index + 1 < speechSynthesizers.count {
```

Comment:
Each time a speech synthesizer fails, we search through the list of speech synthesizers for the index of the one that failed, then turn to the one at the next index. This opens the door to certain classes of developer error. For example, if the developer accidentally specifies the same MySpeechSynthesizer twice in the `speechSynthesizers` list and this speech synthesizer fails for some reason, this method recurses infinitely.

This is pretty unfortunate, but I guess it’s a consequence of each speech synthesizer being asynchronous; all we’re trying to do here is to parallelize them. It does seem to do the right thing if the application makes the controller speak another utterance before it has finished going through the list for the first utterance.

Reply:
I've added a check to `speechSynthesizers` setter to exclude such case.